tokio/src/task/builder.rs

```diff
@@ -0,0 +1,105 @@
+#![allow(unreachable_pub)]
+use crate::util::error::CONTEXT_MISSING_ERROR;
+use crate::{runtime::context, task::JoinHandle};
+use std::future::Future;
+
+/// Factory which is used to configure the properties of a new task.
+///
+/// Methods can be chained in order to configure it.
+///
+/// Currently, there is only one configuration option:
+///
+/// - [`name`], which specifies an associated name for
+///   the task
+///
+/// There are three types of task that can be spawned from a Builder:
+/// - [`spawn_local`] for executing futures on the current thread
+/// - [`spawn`] for executing [`Send`] futures on the runtime
+/// - [`spawn_blocking`] for executing blocking code in the
+///   blocking thread pool.
+///
+/// ## Example
+///
+/// ```no_run
+/// use tokio::net::{TcpListener, TcpStream};
+///
+/// use std::io;
+///
+/// async fn process(socket: TcpStream) {
+///     // ...
+/// # drop(socket);
+/// }
+///
+/// #[tokio::main]
+/// async fn main() -> io::Result<()> {
+///     let listener = TcpListener::bind("127.0.0.1:8080").await?;
+///
+///     loop {
+///         let (socket, socket) = listener.accept().await?;
+///
+///         tokio::task::Builder::new()
+///             .name(&format!("connection from {:?}", socket.ip()))
```

Comment:
i'm not sure if we should encourage people to put formatted strings here --- a better solution would be to eventually allow dynamic sets of fields (which is something future changes to `tracing` will make it easier to support). IMO we'd generally want all instances of a task spawned in the same place to have one `name`, and we'll eventually be able to add things like the IP address as dynamic fields on an instance of a task.

Reply:
Do we want to enforce that by requiring a `&'static str` for name? I intentionally gave the builder a non-'static lifetime to support this dynamic use case, but it sounds like we want to discourage that. Is this a "support it but don't do it in the example" situation, or "it's a footgun, disallow it"

Reply:
i'm somewhat leaning "support it but don't do it in the example" on this one.

but, if we did decide that we want to allow dynamic content here, it occurs to me that the most efficient way to do so would actually be to take an `&impl fmt::Display` or something. `tracing` will record the name as a `Display` trait object, and the value will get written to strings owned by the tracing subscriber, so my main goal is to avoid allocating an intermediate string here --- we don't _need_ that. this would allow passing a `format_args!("blah blah {:?} something dynamic {}", stuff, whatever)` if you wanted dynamically formatted content without allocating.

i could be convinced either way on this, but i think the best idiom is still for names to be fixed across all instances of a task, and to add additional structured fields later. but, it would be good to get @carllerche's opinion?

Reply:
`&impl Display` makes sense as a nice zero-cost way of supporting this.

Relatedly, is there a standard empty placeholder in tracing? This PR currently uses "", assuming that Valuable will add a better way to express a missing value

Reply:
in the Tokio API, I don't think we want to take a `&impl Display`. I do think the usual case **should** be a `&'static str`, but I don't know if we want to enforce it either.

I say, for now, we just take `&str` as it is the simplest and "most obvious" way to take names and move on (it is an unstable API). We can change it later if we get data to support it.

Reply:
Taking an `&impl Display` seems unidiomatic to me, though I understand the point with using it in `format!`.